cmd/tailscale/cli/serve.go

```diff
@@ -115,31 +229,457 @@ func (e *serveEnv) runServe(ctx context.Context, args []string) error {
+			// dir mount points must end in /
+			// for relative file links to work
+			mount += "/"
+		}
+		h.Path = args[2]
+	case "proxy":
+		t, err := expandProxyTarget(args[2])
+		if err != nil {
+			return err
+		}
+		h.Proxy = t
+	case "text":
+		h.Text = args[2]
+	default:
+		fmt.Fprintf(os.Stderr, "error: unknown serve type %q\n\n", args[1])
+		return flag.ErrHelp
+	}
+
+	cursc, err := e.getServeConfig(ctx)
+	if err != nil {
+		return err
+	}
+	sc := cursc.Clone() // nil if no config
+	if sc == nil {
+		sc = new(ipn.ServeConfig)
+	}
+	dnsName, err := e.getSelfDNSName(ctx)
+	if err != nil {
+		return err
+	}
+	hp := ipn.HostPort(net.JoinHostPort(dnsName, srvPortStr))
+
+	if isTCPForwardingOnPort(sc, srvPort) {
+		fmt.Fprintf(os.Stderr, "error: cannot serve web; already serving TCP\n")
+		return flag.ErrHelp
+	}
+
+	mak.Set(&sc.TCP, srvPort, &ipn.TCPPortHandler{HTTPS: true})
+
+	if _, ok := sc.Web[hp]; !ok {
+		mak.Set(&sc.Web, hp, new(ipn.WebServerConfig))
+	}
+	mak.Set(&sc.Web[hp].Handlers, mount, h)
+
+	for k, v := range sc.Web[hp].Handlers {
+		if v == h {
+			continue
+		}
+		// If the new mount point ends in / and another mount point
+		// shares the same prefix, remove the other handler.
+		// (e.g. /foo/ overwrites /foo)
+		// The opposite example is also handled.
+		if (strings.HasSuffix(mount, "/") && k == mount[:len(mount)-1]) ||
+			(strings.HasSuffix(k, "/") && mount == k[:len(k)-1]) {
+			delete(sc.Web[hp].Handlers, k)
+			continue
+		}
+	}
+
+	if !reflect.DeepEqual(cursc, sc) {
+		if err := e.setServeConfig(ctx, sc); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *serveEnv) handleWebServeRemove(ctx context.Context, mount string) error {
+	srvPort, srvPortStr, err := e.validServePortWithString()
+	if err != nil {
+		return err
+	}
+	sc, err := e.getServeConfig(ctx)
+	if err != nil {
+		return err
+	}
+	if sc == nil {
+		return errors.New("error: serve config does not exist")
+	}
+	dnsName, err := e.getSelfDNSName(ctx)
+	if err != nil {
+		return err
+	}
+	if isTCPForwardingOnPort(sc, srvPort) {
+		return errors.New("cannot remove web handler; currently serving TCP")
+	}
+	hp := ipn.HostPort(net.JoinHostPort(dnsName, srvPortStr))
+	if !httpHandlerExists(sc, hp, mount) {
+		return errors.New("error: serve config does not exist")
+	}
+	// delete existing handler, then cascade delete if empty
+	delete(sc.Web[hp].Handlers, mount)
+	if len(sc.Web[hp].Handlers) == 0 {
+		delete(sc.Web, hp)
+		delete(sc.TCP, srvPort)
+	}
+	// clear empty maps mostly for testing
+	if len(sc.Web) == 0 {
+		sc.Web = nil
+	}
+	if len(sc.TCP) == 0 {
+		sc.TCP = nil
+	}
+	if err := e.setServeConfig(ctx, sc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func httpHandlerExists(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) bool {
+	_, ok := getHTTPHandler(sc, hp, mount)
+	return ok
+}
+
+func getHTTPHandler(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) (handler *ipn.HTTPHandler, ok bool) {
```

Comment:
you could have this just return nil when it doesn't exist

```
func httpHandlerExists(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) bool {
	return getHTTPHandler(sc, hp, mount) != nil
}

// getHTTPHandler returns the HTTPHandler for the hp and mount. It returns
// nil if not present.
func getHTTPHandler(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) *ipn.HTTPHandler {
	if sc.Web[hp] != nil {
		return sc.Web[hp].Handlers[mount]
	}
	return nil
}
```

looking up from a `nil` map is fine.

Reply:
oof yeah should've been rolled up... done but also also checking `sc != nil`